packages/qdrant-loader/src/qdrant_loader/core/pipeline/workers/upsert_worker.py

```
@@ -310,10 +332,31 @@ async def process_embedded_chunks(
+                ``(parent_document, success)`` the moment a document's last
+                chunk is accounted for — before the whole iterator finishes.
+                Lets callers persist document state incrementally, so a
+                mid-run interruption doesn't lose the state of documents
+                that were already fully upserted. Callback failures are
+                logged and never abort the pipeline.
 
         Returns:
             PipelineResult with processing statistics
         """
+
+        async def _notify(parent_doc: Any, success: bool) -> None:
```

Ruff (0.15.21)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `parent_doc`

(ANN401)